File: package/main/src/Function/memoize.ts
/**
 * Options for the memoize function.
 */
export interface MemoizeOptions<K> {
  /** Maximum number of entries to store in the cache. */
  maxSize?: number;
  /** Custom function to generate cache keys from arguments. */
  resolver?: (...arguments_: unknown[]) => K;
}

/**
 * A memoized function with an exposed cache.
 */
export interface MemoizedFunction<A extends unknown[], R, K = unknown> {
  (...arguments_: A): R;
  /** The underlying cache Map. */
  cache: Map<K, R>;
}

/**
 * Creates a memoized version of the provided function. Results are
 * cached in a Map keyed by the first argument (or a custom resolver).
 *
 * @param function_ - The function to memoize.
 * @param options - Configuration for cache behavior.
 * @param options.maxSize - Maximum cache entries before evicting the oldest.
 * @param options.resolver - Custom key resolver function.
 * @returns The memoized function with an exposed cache property.
 *
 * @example
 * ```typescript
 * const memoized = memoize((n: number) => n * 2);
 * memoized(5); // 10 (computed)
 * memoized(5); // 10 (cached)
 * memoized.cache.size; // 1
 * ```
 */
export const memoize = <A extends unknown[], R, K = unknown>(
  function_: (...arguments_: A) => R,
  options: MemoizeOptions<K> = {},
): MemoizedFunction<A, R, K> => {
  const { maxSize, resolver } = options;
  const cache = new Map<K, R>();

  const memoized = (
    resolver
      ? function (this: unknown, ...arguments_: A): R {
          const key = resolver(...arguments_);
          const cached = cache.get(key);
          if (cached !== undefined) {
            return cached;
          }
          const result = function_.apply(this, arguments_);
          if (maxSize !== undefined && cache.size >= maxSize) {
            const firstKey = cache.keys().next().value as K;
            cache.delete(firstKey);
          }
          cache.set(key, result);
          return result;
        }
      : function (this: unknown, ...arguments_: A): R {
          const key = arguments_[0] as K;
          const cached = cache.get(key);
          if (cached !== undefined) {
            return cached;
          }
          const result = function_.apply(this, arguments_);
          if (maxSize !== undefined && cache.size >= maxSize) {
            const firstKey = cache.keys().next().value as K;
            cache.delete(firstKey);
          }
          cache.set(key, result);
          return result;
        }
  ) as MemoizedFunction<A, R, K>;

  memoized.cache = cache;

  return memoized;
};

File: package/umt_python/src/function/memoize.py
from collections import OrderedDict
from collections.abc import Callable
from typing import Any


class MemoizedFunction:
    """
    A memoized function with an exposed cache.

    The cache is an OrderedDict that maps keys to computed results.
    When called, the memoized function checks the cache before computing.
    """

    def __init__(
        self,
        fn: Callable[..., Any],
        max_size: int | None = None,
        resolver: Callable[..., Any] | None = None,
    ) -> None:
        self._fn = fn
        self._max_size = max_size
        self._resolver = resolver
        self.cache: OrderedDict[Any, Any] = OrderedDict()

    def __call__(self, *args: Any) -> Any:  # noqa: ANN401
        key = self._resolver(*args) if self._resolver is not None else args[0]

        if key in self.cache:
            return self.cache[key]

        result = self._fn(*args)

        if self._max_size is not None and len(self.cache) >= self._max_size:
            self.cache.popitem(last=False)

        self.cache[key] = result
        return result


def memoize(
    fn: Callable[..., Any],
    *,
    max_size: int | None = None,
    resolver: Callable[..., Any] | None = None,
) -> MemoizedFunction:
    """
    Creates a memoized version of the provided function.

    Results are cached and keyed by the first argument by default,
    or by a custom resolver function if provided.

    Args:
        fn: The function to memoize.
        max_size: Maximum cache entries before evicting the oldest.
        resolver: Custom key resolver function.

    Returns:
        A MemoizedFunction with an exposed cache attribute.

    Example:
        >>> memoized = memoize(lambda n: n * 2)
        >>> memoized(5)
        10
        >>> memoized(5)
        10
        >>> len(memoized.cache)
        1
    """
    return MemoizedFunction(fn, max_size=max_size, resolver=resolver)
